SDL support

alexytomi opened the pull request:
~~A clone of https://github.com/PojavLauncherTeam/PojavLauncher/pull/6462, but Amethyst~~

Not so much anymore

This PR basically takes only the minimum out of what's needed to get controller pass-through working properly, anything else wasn't given thought.

Thanks isXander for most of the reference material used for this PR

alexytomi commented:
This compiled locally wtf

ThatStella7922 commented:
let's see

EDIT 1: Looks sane at first glance, will test on the following physical devices shortly.

Android device:
```
Vendor: OnePlus Co Ltd.
Model: OPD2403 (OnePlus Pad 2)
Android: 15 (OxygenOS 15, September patch)
SoC: Qualcomm Snapdragon 8 Gen 3
Graphics: Qualcomm Adreno 750 (Vk 1.3/GLES 3.2)
```

Controller:
```
Vendor: Microsoft Corporation
Model: 1708 (Xbox Wireless Controller)
Connection method: Bluetooth
```

EDIT 2: it might be a sec, `Permission controller` (yes that one) is broken and I cannot grant permissions to anything

ThatStella7922 commented:
Finally testing after I had to re-flash and reset the pad 2.

Artifact url: https://github.com/AngelAuraMC/Amethyst-Android/actions/runs/17681687420
Artifact: `app-debug`

## Process:
1. Connect controller over BT (controller works to navigate android)
2. Open app
3. Login
4. Create fabric 1.21.8 profile with default settings, add in MidnightControls 1.11.1
5. Enable sdl controller backend in settings. <img width="908" height="69" alt="image" src="https://github.com/user-attachments/assets/93a03784-eaf0-4744-b002-e6ea94704416" />
7. press play

## Results:
Functionality: Does not work to control the game at all.
MidnightControls controller list:
<img width="1078" height="153" alt="image" src="https://github.com/user-attachments/assets/84b47dfb-26cf-4c59-994f-6b140d8d4a34" />
Only shows the emulated Pojav gamepad, the Xbox Wireless Controller does not show up whatsoever.


## Other notes while testing:
<img width="1055" height="181" alt="image" src="https://github.com/user-attachments/assets/86bf9b6f-be3c-4ee6-accd-7225d5620237" />
Launcher sits here for multiple minutes before actually launching game, even if force quit and relaunched. Could very well be internet oddity on my end so ????? ig if this doesn't happen for you its fine.

alexytomi commented:
Hoo, I was using Controlify and Legacy4J for testing, I am able to reproduce the MidnightControls issue.

alexytomi commented:
~~AH, MidnightControls uses SDL2, that has not been added.~~

No it doesn't, its GLFW.

ThatStella7922 commented:
Important update: Controlify works flawlessly because it does use SDL3.

My new verdict is that this appears to work well. I have only attached one controller, but it works.

Sorry about the errant close and reopen, that was a misclick.

ThatStella7922 commented:
<img width="3000" height="2120" alt="Screenshot_20250913-122914" src="https://github.com/user-attachments/assets/3bc0b920-9544-4797-b97f-43f1e5c14c3c" />
Works fine after the change, however the glfw controller setup wizard is shown now

alexytomi commented:
Appears to be caused by both Legacy4J using GLFW forcibly on android and the now present call to `CallbackBridge.enableGamepadDirectInput()` which was missing before, which means GLFW wasn't being properly loaded then.

This needs some more testing, perhaps we can find a way to get rid of the scrolling issues by using this, when I tested the build, it let the screen work properly, however that bug isn't really consistent so it may be a fluke.

alexytomi commented:
Nope, trying to fully kill off GLFW causes infinite scrolling, probably write a hook to whatever triggers the HIDDevice polling to also trigger the SDL capture while keeping the virtual gamepad.

alexytomi commented:
I've now modified it so that it waits till the mod initializes the joystick subsystem before swapping over to the SDL handling, similar to GLFW. I haven't found a controller mod using SDL3 that does not initialize this, I don't see why you wouldn't either.

To do so, I use `SDLControllerManager.pollInputDevices()` which is called by JNI repeatedly as seen here
https://github.com/libsdl-org/SDL/blob/718034f5fa6e796cf609db32ee8d702fc57d78aa/src/joystick/android/SDL_sysjoystick.c#L473

Yes this is jank, we should really be using a hook into `SDL_Init` but I can't figure out how to callback from there so /shrug

Also added sdl2-compat. We didn't need a fork thanks to the SDL guys allowing a PR to upstream instead!

ThatStella7922 commented:
Tested https://github.com/AngelAuraMC/Amethyst-Android/actions/runs/17774945489 (commit cc64450)

# It works

With a caveat.
This is how it works with "Enable SDL controller backend"
<img width="3000" height="2120" alt="Screenshot_20250916-230312" src="https://github.com/user-attachments/assets/117ae298-98de-4f46-aa07-af90e5bffea8" />
You can see it just doesn't initialize the SDL subsystem apparently. No controller detected, glfw controller works.

With "Enable SDL controller backend" and "Forcibly start SDL controller handling"
<img width="3000" height="2120" alt="Screenshot_20250916-230409" src="https://github.com/user-attachments/assets/5fd92f15-f774-4142-862d-b82b9949ac37" />
Works.

alexytomi commented:
This is intended, the l4j android detection fix was very recent. As in only the latest action build properly autoselects SDL instead of loading GLFW.

I can try trigerring force enable if l4j is detected and log says it selected GLFW. This won't fix the auto detection for GLFW so they'll still need to manually do that for the current rels

alexytomi commented:
# Changes
- "Enable SDL controller backend"
    - Enables SDL
    - Input passthrough to our keys input forwarder is blocked when SDL is active as it causes issues with duplicate inputs.
    - GLFW will still be present be cause there is no way to remove or disable it, if we do so forcefully, even more buggier behavior appears. To minimize the effect of this buggy behavior, I have disabled input passthrough to GLFW when the SDL subsystems are active. 
       - Test: If disabled, we should have identical behavior as to before this PR with all inputs, controller or not.

- "Forcibly start SDL controller handling"
    - Starts the SDL subsystems for gamepad, joystick, and events in`MainActivity.onCreate()`.

- Legacy4J old version detection
    - Detects whether or not Legacy4J properly loaded our SDL or used GLFW
        - If GLFW was used
            - Start the SDL subsystems for gamepad, joystick, and events.
            - Show a warning to enable the SDL backend for best experience
    - Technically this is bugged in versions 1.7.4 and below as it never removes the log listener, in that case but I am not supporting that. Update already or deal with the immense overhead of essentially having the log output open at all times (ooo scary, not really tbh).
        - We should use regex to filter out 1.7.5 and above but I don't think anyone is gonna be using those ancient versions of Legacy4J any time soon so am not gonna

- Controllable native loading mitigation
    - If controllable is found, we try to delete the natives it extracts asap, before it can dlopen them and crash because it's dlopening a glibc lib in android
        - This is a horrible race condition, a bug report should be filed with controllable to fix this.
        - This happens because controllable sets the `setExtractionPath()` which only falls back if the path isn't valid, else it crashes trying to dlopen the linux glibc native

ThatStella7922 commented:
Testing https://github.com/AngelAuraMC/Amethyst-Android/actions/runs/17873979455

Minecraft 1.21.8 fabric loader 0.17.2

Vanilla game:
![Screenshot_2025-09-21-00-42-29-59_59e3738f718bc4374df4f834029b3543](https://github.com/user-attachments/assets/e3ad9590-eed2-462f-acb0-9d56f8371e75)
Works fine, glfw controller setup wizard appeared and glfw controller support works

Controllify 2.4.1:
![Screenshot_2025-09-21-00-49-19-50_59e3738f718bc4374df4f834029b3543](https://github.com/user-attachments/assets/0670eb0f-b254-4079-9f6f-a6c3dbc26dc2)
Worked fine too, glfw controller support disabled itself and it was purely SDL. No more pojav cursor 

I think it's good ngl, this gets the Mommy Stella Seal of Approval™️

alexytomi commented:
> Worked fine too, glfw controller support disabled itself and it was purely SDL. No more pojav cursor
GLFW wasn't actually turned off, Controlify turns off their GLFW backend when SDL is found.

It can still be selected in Legacy4J but the inputs won't work because of me adding an if statement to not forward inputs into GLFW when SDL joystick subsystem is running.

alexytomi commented:
# Bugfixes
- The SDL callback did not run only once as intended
- Accidentally blocked all non-controller inputs when SDL enabled

alexytomi commented:
Also just for documentation purposes, the Gamepad remapping system wasn't used because I don't wanna do the job that SDL already does, that is filter all that stuff through to an actual input again.

`eventKeycode >= KeyEvent.KEYCODE_BUTTON_A && eventKeycode <= KeyEvent.KEYCODE_BUTTON_MODE` ensures no weird shit controllers will crash SDL, like an oculus one

If someone somehow has a weirdshit controller they wanna use that isn't registering because of this, they can feel free to make a bug report and I'll try to fix it.

alexytomi commented:
# Bugfixes
- Android's garbage keyevents was leaking through to processKeyEvent again due to fixing up the keyboard input blocking
- Controllable was at the mercy of whether inotify would want to work, now we just poll it..a lot, say hi to fried CPU
